add task counts to HighLevelGraph and Layer html reprs

kori73 opened the pull request:
- [x] Closes #8497
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

I am simply passing the len. Is this okay?

Example:

```
arr = da.ones(shape=(100, 100), chunks=[5,10])
arr2 = da.ones(shape=(100, 100))
res = arr.dot(arr2)
```

![Screenshot from 2022-01-20 00-47-36](https://user-images.githubusercontent.com/36677585/150219468-e4b03f3f-2c4f-4f2f-84b9-6413a2b5a837.png)

GPUtester commented:
Can one of the admins verify this patch?

quasiben commented:
add to allowlist

jsignell commented on dask/highlevelgraph.py:
```diff
@@ -511,6 +511,7 @@ def layer_info_dict(self):
         info = {
             "layer_type": type(self).__name__,
             "is_materialized": self.is_materialized(),
+            "tasks (unoptimized)": f"{len(self)}",
```
I think this will be number of layers not number of tasks. Right @ian-r-rose?

ian-r-rose replied:
I believe this is `Layer.__len__`, not `HighLevelGraph.__len__`, so it should be the right length for this part.

ian-r-rose reviewed (commented):
Thanks @kori73! I actually think that calling `len()` on HighLevelGraph Layers is not quite as safe as we would hope.

With the current design, `__len__()` gives the length of the *materialized* graph. But a lot of the time we don't actually know that length until it is materialized, as specific culling operations can dramatically change the length. So calling `__len__()` canl force materialization before we are ready.

See, e.g., the implementation for the `BroadcastJoinLayer`, which constructs the full task graph upon calling `__len__()`: 
https://github.com/dask/dask/blob/4228dc7439f4ecd05708641bbe25adbf4d658f92/dask/layers.py#L919-L936


The thing that *should* be safe to call is `get_output_keys()`, which is explicitly meant to not force materialization:
https://github.com/dask/dask/blob/4228dc7439f4ecd05708641bbe25adbf4d658f92/dask/highlevelgraph.py#L94-L109

This has a slightly different meaning from the materialized graph length, in that it is restricted only to keys that might be of interest to other layers, rather than intermediate results or other private-ish tasks.

So I'd recommend tweaking the goal of this PR to either:

1. Only show `__len__` if the graph is materialized.
2. Show the length of `get_output_keys()` instead (and maybe change the label from `tasks (unoptimized)` to something like `number of outputs`.

I'd probably prefer the second, as it's closer to the heart of what HighLevelGraphs are meant to do, which is reason about inputs/outputs without forcing a full graph construction.

kori73 commented:
Thanks a lot for the detailed explanation @ian-r-rose! I was suspecting that this might not be the right approcach without knowing why exactly (was thinking that this might force some unwanted computation). I will update the PR according to your suggestion.

kori73 commented:
I have updated to use:

for layers: `get_output_keys`
for graph: `get_all_external_keys` which seems to call `get_output_keys` for each layer.

Updated example:

![Screenshot from 2022-01-22 18-38-45](https://user-images.githubusercontent.com/36677585/150645343-2981d672-3ee4-4036-a91d-6a7302904dd7.png)

bryanwweber commented:
@ian-r-rose Can you take another look here when you get a chance?

ian-r-rose commented on dask/widgets/templates/highlevelgraph.html.j2:
```diff
@@ -15,7 +15,7 @@
             <h3 style="margin-bottom: 0px;">HighLevelGraph</h3>
             <p style="color: var(--jp-ui-font-color2, #5D5851); margin-bottom:0px;">
-                {{ type }} with {{ layers | length }} layers.
+                {{ type }} with {{ layers | length }} layers and {{ n_outputs }} outputs.
```
I'm finding the term "outputs" to be a bit ambiguous. As used here, it's all the "output" keys for each individual layer. But I could also imagine it being meaning the number of output keys for the *last* layer. That is to say, the number of outputs for the whole computation, discarding intermediate keys.

We could, of course, include both. But since users are also able to inspect the number of outputs in the last layer already, maybe we could just say something like "`and {{ n_outputs }} keys from all layers`". Or maybe it's okay as-is. What do you think @kori73 ?

kori73 replied:
I totally agree that `outputs` sounds like the output keys for the last layer. `keys from all layers` would be much less ambiguous.

ian-r-rose reviewed (commented):
So sorry for the slow review @kori73! I'm happy with this implementation as it is, I just have some musings about how best to describe these keys, which I think are probably confusing to a lot of users. I'd be curious to hear your thoughts.

ian-r-rose commented:
> I also have another concern unrelated to this discussion. I just want to make sure that we are not doing anything that will harm the user experience. According to #8570 `get_all_external_keys` could cause memory problems (i.e. generation of billions of keys).

At least for `get_all_external_keys()`, the set is cached, so it should be safe to call it more than once: https://github.com/dask/dask/blob/bd8e8dc3f1329e7f9ef24072479729d9effeb6ad/dask/highlevelgraph.py#L772

I would say that the *intention* of `get_all_external_keys()` is that it is fairly cheap to call. So using it here would not be a misuse of the API. But, as you point out, there are some issues with the design of HighLevelGraph algorithms that make it not as cheap to manipulate as intended. To my mind, that's something that will need to be addressed in the short term, though I don't think we need to do it here. Any refactoring of the highlevelgraph cull operation will need to have *something* that allows you to reason about how "big" each graph is, and it should be easy to update the repr here should we need to.

TL;DR, I think this is fine, and would probably be easy to update if we need to

ian-r-rose reviewed (approved):
Thanks @kori73!
